packages/angular_devkit/core/src/jobs/README.md

```diff
@@ -0,0 +1,386 @@
+dispatcher.addConditionalDelegate(() => userWantsYarn, yarnInstall.name);
+```
+
+## Execution Strategy
+Jobs are always run in parallel and will always start, but many helper functions are provided 
+when creating a job to help you control the execution strategy;
+
+1. `serialize()`. Multiple runs of this job will be queued with each others.
+1. `memoize(replayEvents = false)` will create a job, or reuse the same job  when inputs are 
+matching. If the inputs don't match, a new job will be started and its outputs will be stored.
+
+These strategies can be used when creating the job:
+
+```typescript
+// Same input and output as above.
+
+export const add = jobs.strategy.memoize()(
+  jobs.createJob<number[], number>(
+      input => input.reduce((total, curr) => total + curr, 0),
+  ),
+);
+```
+
+Strategies can be reused to synchronize between jobs. For example, given jobs `jobA` and `jobB`, 
+you can reuse the strategy to serialize both jobs together;
+
+```typescript
+const strategy = jobs.strategy.serialize();
+const jobA = strategy(jobs.createJob(...));
+const jobB = strategy(jobs.createJob(...));
+```
+
+Even further, we can have package A and package B run in serialization, and B and C also be 
+serialized. Running A and C will run in parallel, while running B will wait for both A and C 
+to finish.
+
+```typescript
+const strategy1 = jobs.strategy.serialize();
+const strategy2 = jobs.strategy.serialize();
+const jobA = strategy1(jobs.createJob(...));
+const jobB = strategy1(strategy2(jobs.createJob(...)));
+const jobC = strategy2(jobs.createJob(...));
+```
+
+# Scheduling Jobs
+Jobs can be scheduled using a `Scheduler` interface, which contains a `schedule()` method:
+
+```typescript
+interface Scheduler {
+  /**
+   * Schedule a job to be run, using its name.
+   * @param name The name of job to be run.
+   * @param argument The argument to send to the job when starting it.
+   * @param options Scheduling options.
+   * @returns The Job being run.
+   */
+  schedule<I extends MinimumInputValueT, O extends MinimumOutputValueT>(
+    name: JobName,
+    argument: I,
+    options?: ScheduleJobOptions,
+  ): Job<JsonValue, O>;
+}
+```
+
+The scheduler also has a `getDescription()` method to get a `JobDescription` object for a certain 
+name; that description contains schemas for the argument, input, output, and other channels:
+
+```typescript
+interface Scheduler {
+  /**
+   * Get a job description for a named job.
+   *
+   * @param name The name of the job.
+   * @returns A description, or null if the job cannot be scheduled.
+   */
+  getDescription(name: JobName): JobDescription | null;
+    
+  /**
+   * Returns true if the job name has been registered.
+   * @param name The name of the job.
+   * @returns True if the job exists, false otherwise.
+   */
+  has(name: JobName): boolean;
+}
+```
+
+Finally, the scheduler interface has a `pause()` method to stop scheduling. This will queue all 
```

Comment:
They're executed normally. This only prevent jobs in the `queued` state from running.

Reply:
Actually now that I'm thinking about it, maybe it should prevent jobs from being `Created`, but still `Queued`. Essentially doing all the work except actually running the job. WDYT?

Reply:
I think allowing jobs to be queued is better. Because a running job can actually queue new jobs using `JobHandlerContext`. So if queuing is prevented, a running job could fail/crash unexpectedly because the scheduler is paused.

Looking at the states from the pause point of view I see `Created` as meaning `Ready`. A job can be ready to run but awaiting to be scheduled. I don't know if this is really a better name but that helped me understand the difference between `Created` and `Started`.

Reply:
what about a "schedule then execute" pattern?  That would probably cover both major usage scenarios: schematic tasks and architect.

Reply:
Jobs are lazy, and will not start until they're subscribed to. I think that covers the "schedule then execute pattern" for the scheduling part. What I'm talking about is the job implementation side, where you might want to synchronize using complex logic. That's done through strategies.

Reply:
> Created is really past the point where we are waiting for the scheduler to resume.

I feel the more we talk about the lifecycle the more confusing the the `Created` state is :/

- Jobs are created through `jobs.createJob`, but that really has nothing to do with the `Created` state. 
- What's the situation where it matters that something is `Created` but not `Started`? 
- Can anything stop that transition from `Created` to `Started` aside from argument validation? 
- Why is argument validation happening during the `Created` state when the argument and schema are known when scheduling a job?

Reply:
If the jobs don't start until the output is subscribed then `schedule` is really a misnomer.  The job consumer is really the scheduler as it decides when to subscribe.

Reply:
> Jobs are created through jobs.createJob, but that really has nothing to do with the Created state.

You're confusing running state with API factory function.

> What's the situation where it matters that something is Created but not Started?

Argument validation fails. Job won't start. Also, `Started` is the responsibility of the Job itself.

> Can anything stop that transition from Created to Started aside from argument validation?

The Job could never emit `Started`.

> Why is argument validation happening during the Created state when the argument and schema are known when scheduling a job?

That's a good question and I'm opened to remove the `Created` state (replacing it with `Started`), move validation before dependencies and remove the Job's responsibility to have its own state management.

----

> If the jobs don't start until the output is subscribed then schedule is really a misnomer. The job consumer is really the scheduler as it decides when to subscribe.

@clydin Please move unrelated discussion to a separate thread. This is already becoming hard to follow.

Reply:
It's all related to the lifecycle of a job though.  Queued/Created/Started -- when do they happen and what controls the progression.
In regards to Filipe's comment: is `Created` but not `Started` tied to the subscription of the output for instance or will the job never get passed `Queued` in that situation?

Reply:
It's not that I was confusing `jobs.createJob` with the `Created` state, but rather that I wanted to highlight how I felt that added to the confusion.

I don't have an opinion regarding what entity should manage a Jobs state really. Just think that the `Created` state is odd and confusing. 

I can imagine a use case for a state before `Started`: once a job is ready to run but the scheduler is preventing existing jobs from starting due to resource contention. 

There are probably other situations where a state before `Started` matters. It just doesn't feel like validating arguments is one.

Reply:
Renamed the `Created` event to `Ready`. Would that be a more reasonable name?

Reply:
I think it makes it easier to understand, yes. Thank you.